### PR TITLE
[Merged by Bors] - refactor(FieldTheory/PerfectClosure): change the order of some results

### PR DESCRIPTION
**Pull request by acmepjz**
- Now the perfect closure is a perfect ring without requiring base ring to be a field.
- Add `mk_pow` and `mk_surjective`.
- Add explicit names to all of the instances.
- Add docstrings for the file.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by riccardobrasca — APPROVED** [review]
Since you are at it, do you mind add a short section at the beginning with a description of the content of the file? Thanks!

bors d+

**Comment by acmepjz** [comment]
> Since you are at it, do you mind add a short section at the beginning with a description of the content of the file? Thanks!

Docstrings added. Does it look OK?

File: Mathlib/FieldTheory/PerfectClosure.lean [hunk 1]
```diff
@@ -8,7 +8,41 @@ import Mathlib.FieldTheory.Perfect
...
+# The perfect closure of a characteristic `p` ring
+
+## Main definitions
+
+- `PerfectClosure`: the perfect closure of a characteristic `p` ring, which is the smallest
+  extension that makes frobenius surjective.
+
+- `PerfectClosure.mk K p (n, x)` for `n : ℕ` and `x : K` is an element of `PerfectClosure K p`,
+  viewed as `x ^ (p ^ -n)`. Every element of `PerfectClosure K p` is of this form
+  (`PerfectClosure.mk_surjective`).
```
Comment (riccardobrasca) [inline Mathlib/FieldTheory/PerfectClosure.lean:21]:
```suggestion
- `PerfectClosure.mk K p (n, x)`: for `n : ℕ` and `x : K` this is `x ^ (p ^ -n)` viewed as an element of `PerfectClosure K p`. Every element of `PerfectClosure K p` is of this form
  (`PerfectClosure.mk_surjective`).
```

File: Mathlib/FieldTheory/PerfectClosure.lean [hunk 2]
```diff
@@ -8,7 +8,41 @@ import Mathlib.FieldTheory.Perfect
...
+
+- `PerfectClosure.of`: the structure map from `K` to `PerfectClosure K p`.
+
+- `PerfectClosure.lift`: given a ring `K` of characteristic `p` and a perfect ring `L` of the same
+  characteristic, any homomorphism `K →+* L` can be lifted to `PerfectClosure K p`.
+
+## Main results
+
+- `PerfectClosure.induction_on`: to prove a result for all elements of the prefect closure, only
+  need to prove it for all `x ^ (p ^ -n)`.
```
Comment (riccardobrasca) [inline Mathlib/FieldTheory/PerfectClosure.lean:31]:
```suggestion
- `PerfectClosure.induction_on`: to prove a result for all elements of the prefect closure, one only
  needs to prove it for all elements of the form `x ^ (p ^ -n)`.
```

File: Mathlib/FieldTheory/PerfectClosure.lean [hunk 3]
```diff
@@ -8,7 +8,41 @@ import Mathlib.FieldTheory.Perfect
...
+
+- `PerfectClosure.mk_mul_mk`, `PerfectClosure.one_def`, `PerfectClosure.mk_add_mk`,
+  `PerfectClosure.neg_mk`, `PerfectClosure.zero_def`, `PerfectClosure.mk_zero_zero`,
+  `PerfectClosure.mk_zero`, `PerfectClosure.mk_inv`, `PerfectClosure.mk_pow`:
+  how to do multiplication, addition, etc. on elements of form `x ^ (p ^ -n)`.
```
Comment (riccardobrasca) [inline Mathlib/FieldTheory/PerfectClosure.lean:36]: I think those are not needed, but as you prefer.

File: Mathlib/FieldTheory/PerfectClosure.lean [hunk 4]
```diff
@@ -8,7 +8,41 @@ import Mathlib.FieldTheory.Perfect
...
+
+- `PerfectClosure.eq_iff'`: when does `x ^ (p ^ -n)` equal.
```
Comment (riccardobrasca) [inline Mathlib/FieldTheory/PerfectClosure.lean:38]: I think we can also ignore it, but the name should be `PerfectClosure.mk_eq_iff`.

**Review by riccardobrasca — APPROVED** [review]
Thanks a lot. Just a couple of cosmetic comments and a name change (not your fault, and hopefully a trivial change).

File: Mathlib/FieldTheory/PerfectClosure.lean (hunk 4 above)
Comment (acmepjz) [reply Mathlib/FieldTheory/PerfectClosure.lean:38]: In fact the existing API confuses me, there is also `PerfectClosure.eq_iff` but which assumes moreover that `K` is reduced (and whose statement uses `Quot.mk` but not `PerfectClosure.mk` --- aren't they defeq ??? very strange), so the criterion becomes simpler, while the `PerfectClosure.eq_iff'` is stated for all rings.

**Pull request closed without merging**